drivers/counter/counter_ll_stm32_rtc.c

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright (c) 2018 Workaround GmbH
+ * Copyright (c) 2018 Allterco Robotics
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Source file for the STM32 RTC driver
+ *
+ */
+
+#include <time.h>
+
+#include <clock_control/stm32_clock_control.h>
+#include <clock_control.h>
+#include <misc/util.h>
+#include <kernel.h>
+#include <soc.h>
+#include <counter.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(counter_rtc_stm32, CONFIG_COUNTER_LOG_LEVEL);
+
+#define T_TIME_OFFSET 2088656896
+
+#if defined(CONFIG_SOC_SERIES_STM32L4X)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_18
+#elif defined(CONFIG_SOC_SERIES_STM32F4X) \
+	|| defined(CONFIG_SOC_SERIES_STM32F3X)	\
+	|| defined(CONFIG_SOC_SERIES_STM32F7X)
+#define RTC_EXTI_LINE	LL_EXTI_LINE_17
+#endif
+
+struct rtc_stm32_config {
+	struct counter_config_info counter_info;
+	struct stm32_pclken pclken;
+	LL_RTC_InitTypeDef ll_rtc_config;
+};
+
+struct rtc_stm32_data {
+	const struct counter_alarm_cfg *alarm_cfg;
+};
+
+
+#define DEV_DATA(dev) ((struct rtc_stm32_data *const)(dev)->driver_data)
+#define DEV_CFG(dev)	\
+((const struct rtc_stm32_config * const)(dev)->config->config_info)
+
+
+static void rtc_stm32_irq_config(struct device *dev);
+
+
+static int rtc_stm32_start(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	LL_RCC_EnableRTC();
+
+	return 0;
+}
+
+
+static int rtc_stm32_stop(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	LL_RCC_DisableRTC();
+
+	return 0;
+}
+
+
+static u32_t rtc_stm32_read(struct device *dev)
+{
+	struct tm now = { 0 };
+	time_t ts;
+	u32_t rtc_date, rtc_time, ticks;
+
+	ARG_UNUSED(dev);
+
+	/* Read time and date registers */
+	rtc_time = LL_RTC_TIME_Get(RTC);
+	rtc_date = LL_RTC_DATE_Get(RTC);
+
+	/* Convert calendar datetime to UNIX timestamp */
+	now.tm_year = __LL_RTC_CONVERT_BCD2BIN(
+					__LL_RTC_GET_YEAR(rtc_date));
+	now.tm_mon = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MONTH(rtc_date));
+	now.tm_mday = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_DAY(rtc_date));
+
+	now.tm_hour = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_HOUR(rtc_time));
+	now.tm_min = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MINUTE(rtc_time));
+	now.tm_sec = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_SECOND(rtc_time));
+
+	ts = mktime(&now);
+
+	/* Return number of seconds since RTC init */
+	ts -= T_TIME_OFFSET;
+
+	ticks = counter_us_to_ticks(dev, (u64_t)(ts * USEC_PER_SEC));
+
+	return ticks;
+}
+
+static int rtc_stm32_set_alarm(struct device *dev, u8_t chan_id,
+				const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct tm alarm_tm;
+	time_t alarm_val;
+	LL_RTC_AlarmTypeDef rtc_alarm;
+
+	u32_t now = rtc_stm32_read(dev);
+	u32_t ticks = alarm_cfg->ticks;
+
+	DEV_DATA(dev)->alarm_cfg = alarm_cfg;
```

Comment:
api changed so that `alarm_cfg` is not persistent anymore. You need to store user data and handler.